lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -2138,6 +2138,89 @@ class DecomposeAtenNumpyTOp : public OpRewritePattern<AtenNumpyTOp> {
 };
 } // namespace
 
+template <typename OpTy>
+static LogicalResult calculateVariance(OpTy op, PatternRewriter &rewriter,
+                                       bool unbiased, int64_t correction) {
+  Location loc = op.getLoc();
+  Value self = op.self();
+  Value dimList = op.dim();
+  Value keepDim = op.keepdim();
+  Type outputType = op.getType();
+  BaseTensorType inputTensorTy = self.getType().cast<BaseTensorType>();
+  if (!inputTensorTy.hasDtype() ||
+      !inputTensorTy.getDtype().isa<mlir::FloatType>()) {
+    return rewriter.notifyMatchFailure(op, "support floating type input only");
+  }
+
+  unsigned inputRank = getTensorRank(self);
+  SmallVector<Value> dimListElements;
+  bool isNoneOrEmpty = true;
+  if (!dimList.getType().template isa<Torch::NoneType>()) {
+    if (!getListConstructElements(dimList, dimListElements))
+      return rewriter.notifyMatchFailure(
+          op, "expect dimList to be constructed from list construct");
+    if (dimListElements.size() != 0 || inputRank == 0)
+      isNoneOrEmpty = false;
+  }
+  if (isNoneOrEmpty) {
+    for (unsigned i = 0; i < inputRank; i++)
+      dimListElements.push_back(rewriter.create<Torch::ConstantIntOp>(
+          loc, rewriter.getI64IntegerAttr(i)));
+    dimList = rewriter.create<PrimListConstructOp>(
+        loc, Torch::ListType::get(Torch::IntType::get(op.getContext())),
+        dimListElements);
+  }
+  Type meanDimResultType = inputTensorTy;
+  for (unsigned i = 0; i < dimListElements.size(); i++)
+    meanDimResultType = computeReductionType(
+        rewriter, op, meanDimResultType.cast<BaseTensorType>(),
+        dimListElements[i],
+        /*keepDim=*/true);
+
+  Value constantNone = rewriter.create<ConstantNoneOp>(loc);
+  Value constantTrue = rewriter.create<ConstantBoolOp>(loc, true);
+  Value meanAlongDims = rewriter.create<AtenMeanDimOp>(
+      loc, meanDimResultType, self, dimList, /*keepDim=*/constantTrue,
+      /*dtype=*/constantNone);
+  Value subMean =
+      createTensorSub(rewriter, loc, inputTensorTy, self, meanAlongDims);
+  Value square = rewriter.create<AtenSquareOp>(loc, inputTensorTy, subMean);
+  if (!unbiased) {
+    rewriter.replaceOpWithNewOp<AtenMeanDimOp>(op, outputType, square, dimList,
+                                               keepDim, /*dtype=*/constantNone);
+    return success();
+  }
+  // Divide the square sum by productDimSize - correction.
+  Value squareSum = rewriter.create<AtenSumDimIntListOp>(
+      loc, outputType, square, dimList, keepDim, /*dtype=*/constantNone);
+
+  // `productDimSize` is product of sizes of dimensions to be reduced.
+  Value constantOne =
+      rewriter.create<Torch::ConstantIntOp>(loc, rewriter.getI64IntegerAttr(1));
+  Value productDimSize = constantOne;
+  for (Value dim : dimListElements) {
+    Value dimSize = rewriter.create<AtenSizeIntOp>(loc, self, dim);
```

Comment:
I would've expected `AtenSizeIntOp` and `AtenMeanDimOp` to handle the case of `dim=[]` rather than creating a special `dimList` in this decomposition, since their python counterparts reduce to a scalar when `dim=[]`. Is this not the case? If so, that should be fixed first.

```python
In [1]: import torch

In [2]: a = torch.rand((1,2,3))

In [3]: torch.ops.aten.sum(a, dim=[])
Out[3]: tensor(2.9630)

In [4]: torch.ops.aten.sum(a, dim=None)
Out[4]: tensor(2.9630)

In [5]: torch.ops.aten.mean(a, dim=[])
Out[5]: tensor(0.4938)
```

Reply:
I think you mean to say `AtenSumDimIntListOp` and `AtenMeanDimOp`. Yes, we need to fix the `AtenSumDimIntListOp`, that will automatically fix the `AtenMeanDimOp`. But we still need to create this special `dimList` for `AtenSizeIntOp`, which expects `dim` to be an `integer` not a `list`. Also, we need to fix the shape calculation for `AtenSumDimIntListOp`, and `AtenMeanDimOp`.